nixos.nix

```
@@ -4,454 +4,184 @@ let
   inherit (lib)
     attrNames
     attrValues
+    mapAttrsToList
     zipAttrsWith
     flatten
     mkAfter
     mkOption
-    mkDefault
     mkIf
     mkMerge
-    mapAttrsToList
     types
     foldl'
     unique
-    concatMap
     concatMapStrings
-    listToAttrs
     escapeShellArg
     escapeShellArgs
     recursiveUpdate
     all
     filter
     filterAttrs
     concatStringsSep
-    concatMapStringsSep
     catAttrs
-    optional
+    optionals
     optionalString
     literalExpression
     elem
-    mapAttrs
     intersectLists
     any
     id
-    head
+    ;
+
+  inherit (types)
+    attrsOf
+    submodule
     ;
 
   inherit (utils)
     escapeSystemdPath
-    fsNeededForBoot
+    pathsNeededForBoot
     ;
 
   inherit (pkgs.callPackage ./lib.nix { })
-    splitPath
     concatPaths
     parentsOf
     duplicates
     ;
 
+  inherit (config.users) users;
+
   cfg = config.environment.persistence;
-  users = config.users.users;
-  allPersistentStoragePaths = zipAttrsWith (_name: flatten) (filter (v: v.enable) (attrValues cfg));
+
+  # All persistent storage path submodule values zipped together into
+  # one set. This includes paths from the Home Manager persistence
+  # module and `users` submodules.
+  allPersistentStoragePaths =
+    let
+      # All enabled system paths
+      nixos = filter (v: v.enable) (attrValues cfg);
+
+      # Get the files and directories from the `users` submodules of
+      # enabled system paths
+      nixosUsers = flatten (map attrValues (catAttrs "users" nixos));
+
+      # Fetch enabled paths from all Home Manager users who have the
+      # persistence module loaded
+      homeManager =
+        let
+          paths = flatten
+            (mapAttrsToList
+              (_name: value:
+                attrValues (value.home.persistence or { }))
+              config.home-manager.users or { });
+        in
+        filter (v: v.enable) paths;
+    in
+    zipAttrsWith (_: v: flatten v) (nixos ++ nixosUsers ++ homeManager);
```

```suggestion
    zipAttrsWith (_: flatten) (nixos ++ nixosUsers ++ homeManager);
```

Could this use [`concatLists`](https://noogle.dev/f/builtins/concatLists) instead of `flatten`?:

```suggestion
    zipAttrsWith (_: concatListsn) (nixos ++ nixosUsers ++ homeManager);
```